sig-cli/design-principles.md

```diff
@@ -0,0 +1,78 @@
+# Kubectl and SIG CLI Design Principles
+
+## Focus
+
+kubectl provides Resource and Resource Config oriented commands
+(as opposed to some other central concepts, such as packaging, integration, etc).
+This includes but is not limited to commands to generate, transform, create,
+update, delete, watch, print, edit, validate and aggregate information
+about Resources and Resource Config.  This functionality may be either
+declarative or imperative.
```

Comment:
This focus appears to have cross over with the ecosystem. If this is talking about the CRUD of configuration files locally (outside of a cluster), with more feature intent than we have now, would it be in a similar space with existing configuration managers?

Reply:
kubectl is not in a similar space with existing configuration managers. 

Kubectl provides a reference implementation for interacting with the API, with relatively low-level building blocks. We've had a long-standing position (https://github.com/kubernetes/kubernetes/issues/12143) that widely used, general-purpose functionality should be implemented eventually in the server. Past examples of functionality moving to the server are Deployment (rolling-update) and garbage collection (reaping). An ongoing example is server-side apply. Apply and strategic merge patch were important mechanisms pioneered by kubectl. OpenAPI-based validation is another example.

We've long intended that kubectl's implementation to be available in library form as well as a command (https://github.com/kubernetes/kubernetes/issues/7311), but that's been harder than expected, and has been preempted by other priorities, such as extracting kubectl from k/k, which is still an eventual goal, as productivity in k/k is low.

The reference implementation demonstrates how to use the API, including strategic merge patch, as well as providing a simple getting started tool and avoiding complexities of documenting the system with just, for instance, curl. It has long had (relatively simple) commands, such as run, for convenience of expected common operations. The other creation commands, especially create secret and create configmap, are in that category, as well. They help (esp. new users) not worry about schema details and yaml indentation. 

kubectl's scope excludes packaging, dependency management, application publishing and discovery, lifecycle hooks, templating, configuration DSLs, and other things that configuration management tools do. And I haven't seen those tools do things, like apply, that kubectl does. 

From the beginning (http://prs.k8s.io/1325), kubectl was intended to provide resource-oriented bulk declarative and imperative operations. 

I wrote more about the kubectl design ethos in this comment:
https://github.com/kubernetes/community/pull/3164#discussion_r251705527

Reply:
Quick comment: The server-side principle has been documented for a long, long time.

This should be merged with the new principles doc:
https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#principles

Another PR is open to move it, along with the other docs in devel.

Reply:
And the core of kustomize, strategic merge patch, IS on the server side.

Identification of resource cross-references is something that has been identified that needs to be published server-side. Other functionality (e.g., policy enforcement) could benefit from it also.

Reply:
On deployment patterns:

Since 1.0 we've had examples of some deployment patterns, such as for canary deployments:
https://kubernetes.io/docs/concepts/cluster-administration/manage-deployment/#canary-deployments

It doesn't preclude other patterns.

Actually, lots of our documentation needs to be updated and expanded. Few contribute to it, other than to document new features. So, yes, Phil should be applauded, for the new kubectl book, for his past work on kubectl documentation (e.g., https://kubernetes.io/docs/concepts/overview/object-management-kubectl/overview/), for the generated API documentation (https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.13/), and other contributions to the documentation, including the overall site design.

Reply:
@pwittrock FWIW, I appreciate the style you're going for and think content with this kind of flow is needed. The question that comes to my mind how much of this content should be in the k8s project?

For example, the docs have a section for building images. If someone is going to create containers to run in Kubernetes this is very much needed. But, neither kubernetes nor kubectl do this. And, there are numerous builders available. So, how does the k8s project choose which builder to privilege by being in these examples? Should we even be in a place to do that?

I can see why SIG Docs likes the style. The choice of gitbook is one I appreciate, too.

Reply:
> So, how does the k8s project choose which builder to privilege by being in these examples? Should we even be in a place to do that?

I don't think what you are describing is what I envisioned for that section.  I updated the book to demonstrate something close to my original thinking - e.g. decisions the user will make when building an image that will impact how they use the tooling - e.g. using a digest and updating the images references vs using *latest* with an imagePull always.

I'd like to see more of this sort of discussion take place.  I ask these sorts of questions to myself, and being able to discuss them with a quorum of stakeholders would ensure we have the authority to make decisions on these matters.  Perhaps we could driving this out of a subproject within sig-docs and invite stakeholders from workloads / apimachinery / kubectl to participate.